node/bft-consensus/tests/basics.rs

```diff
@@ -136,94 +146,68 @@ async fn primary_failures() {
+        deadline!(Duration::from_secs(10), move || {
+            let mut states = states.iter();
+            let first_tx_count = states.next().unwrap().processed_txs.load(Ordering::SeqCst);
+
+            if first_tx_count == tx_count_before_batch {
+                return false;
+            }
+
+            states.map(|state| state.processed_txs.load(Ordering::SeqCst)).all(|tx_count| tx_count == first_tx_count)
+        });
+
+        // Kill one of the consensus instances and shut down the corresponding transaction client.
+        let instance_idx = rng.gen_range(0..NUM_PRIMARIES - i);
+        let instance = running_consensus_instances.swap_remove(instance_idx);
+        for worker_node in &instance.worker_nodes {
+            worker_node.shutdown().await;
         }
-    }
-
-    // Wait for a while to allow the transfers to be processed.
-    tokio::time::sleep(Duration::from_secs(3)).await;
-
-    // Save the current numbers for processed transactions.
-    let tx_counts2 = running_consensus_instances
-        .iter()
-        .map(|rci| rci.state.processed_txs.load(Ordering::SeqCst))
-        .collect::<Vec<_>>();
-
-    // First check: the processed tx counts should have changed, as a single missing primary shouldn't break the consensus.
-    for (count1, count2) in tx_counts1.iter().zip(&tx_counts2) {
-        assert!(count2 > count1);
-    }
-
-    // Kill another one of the primaries and shut down the corresponding transaction client.
-    let instance_idx = rng.gen_range(0..NUM_PRIMARIES - 1);
-    let instance = running_consensus_instances.remove(instance_idx);
-    // FIXME: this shouldn't need to happen in a separate task, but the await hangs otherwise.
-    tokio::spawn(async move {
         instance.primary_node.shutdown().await;
-    });
-    tx_clients.remove(instance_idx);
-
-    // Send another third of the transactions to the workers.
-    for transfer in &transfers[20..] {
-        let transaction: Bytes = bincode::serialize(&transfer).unwrap().into();
-        let tx = TransactionProto { transaction };
-
-        // Submit the transaction to the chosen workers.
-        for tx_client in &mut tx_clients {
-            tx_client.submit_transaction(tx.clone()).await.unwrap();
-        }
+        drop(instance);
+        // This index matches the consensus instance one due to us sorting the clients by the port.
+        tx_clients.swap_remove(instance_idx);
     }
-
-    // Wait for a while to allow the transfers to be processed.
-    tokio::time::sleep(Duration::from_secs(3)).await;
-
-    // Save the current numbers for processed transactions.
-    let tx_counts3 = running_consensus_instances
-        .iter()
-        .map(|rci| rci.state.processed_txs.load(Ordering::SeqCst))
-        .collect::<Vec<_>>();
-
-    // Final check: the processed tx counts should NOT have changed, as another missing primary should break the consensus.
-    assert_eq!(tx_counts3, &tx_counts2[..2]);
```

Comment:
the test conditions have changed - right now the tx counts are only used to ensure that the state is coherent across the board, and the test passes if the `break` is reached

Reply:
else it keeps on running forever?

Reply:
also, if the tx counts don't change, `false` is returned within the `deadline` macro, which would trigger a panic

Reply:
> else it keeps on running forever?
it hasn't happened in my local runs (it always triggered a timeout eventually), but it would panic due to transfer pool depletion :detective: